server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GeoTileGridValuesSourceBuilder.java

```diff
@@ -0,0 +1,144 @@
+        PARSER.declareInt(GeoTileGridValuesSourceBuilder::precision, new ParseField("precision"));
+        CompositeValuesSourceParserHelper.declareValuesSourceFields(PARSER, ValueType.NUMERIC);
+    }
+
+    static GeoTileGridValuesSourceBuilder parse(String name, XContentParser parser) throws IOException {
+        return PARSER.parse(parser, new GeoTileGridValuesSourceBuilder(name), null);
+    }
+
+    private int precision = GeoTileGridAggregationBuilder.DEFAULT_PRECISION;
+
+    GeoTileGridValuesSourceBuilder(String name) {
+        super(name);
+    }
+
+    GeoTileGridValuesSourceBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.precision = in.readInt();
+    }
+
+    public GeoTileGridValuesSourceBuilder precision(int precision) {
+        this.precision = GeoTileUtils.checkPrecisionRange(precision);
+        return this;
+    }
+
+    @Override
+    public GeoTileGridValuesSourceBuilder format(String format) {
+        throw new IllegalArgumentException("[format] is not supported for [" + TYPE + "]");
+    }
+
+    @Override
+    protected void innerWriteTo(StreamOutput out) throws IOException {
+        out.writeInt(precision);
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, Params params) throws IOException {
+        builder.field("precision", precision);
+    }
+
+    @Override
+    String type() {
+        return TYPE;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), precision);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (obj == null || getClass() != obj.getClass()) return false;
+        if (super.equals(obj) == false) return false;
+        GeoTileGridValuesSourceBuilder other = (GeoTileGridValuesSourceBuilder) obj;
+        return precision == other.precision;
+    }
+
+    @Override
+    protected CompositeValuesSourceConfig innerBuild(SearchContext context, ValuesSourceConfig<?> config) throws IOException {
+        ValuesSource orig = config.toValuesSource(context.getQueryShardContext());
+        if (orig == null) {
+            orig = ValuesSource.GeoPoint.EMPTY;
+        }
+        if (orig instanceof ValuesSource.GeoPoint) {
+            ValuesSource.GeoPoint geoPoint = (ValuesSource.GeoPoint) orig;
+            // is specified in the builder.
+            final MappedFieldType fieldType = config.fieldContext() != null ? config.fieldContext().fieldType() : null;
+            CellIdSource cellIdSource = new CellIdSource(geoPoint, precision, GeoTileUtils::longEncode);
+            return new CompositeValuesSourceConfig(name, fieldType, cellIdSource, GEOTILE, order(), missingBucket());
+        } else {
+            throw new IllegalArgumentException("invalid source, expected geo_point, got " + orig.getClass().getSimpleName());
+        }
+    }
+
+    DocValueFormat GEOTILE = new DocValueFormat() {
```

Comment:
Since this is the only place it is used, I put it here. But, we may want to support something like this elsewhere? I am not sure. 

If so, I will happily move it to the docvalue format file. Also, I used the name `geo_tile` as a parallel to `geo_hash` but I am not sure if that is the correct name for it. 

@talevy let me know what you think.

Reply:
ah right. I think that it is fine to place here for the reasons you stated. Would this be nicer to have closer to the top of the file and set to private+static?

Reply:
@talevy updated :)

Reply:
@talevy the `geo_tile` needs to be writable/readable via streams. So, I think moving it to the `DocValueFormat` file may be best. I did such. Let me know what you think.

Reply:
the move makes sense given what Zach mentioned below!